src/audio/volume/volume.c

```diff
@@ -1182,9 +1181,14 @@ static int volume_prepare(struct comp_dev *dev)
 	if (ret == COMP_STATUS_STATE_ALREADY_SET)
 		return PPL_STATUS_PATH_STOP;
 
-	/* volume component will only ever have 1 sink buffer */
+	/* volume component will only ever have 1 sink and source buffer */
 	sinkb = list_first_item(&dev->bsink_list,
 				struct comp_buffer, source_list);
+	sourceb = list_first_item(&dev->bsource_list,
+				  struct comp_buffer, sink_list);
+	/* volume component process need sink and source frames aligned with 4 */
+	sinkb->stream.frame_align_shift = 2;
```

Comment:
Wont this depend on format, channels and HiFi version ? i.e. hard coding it to 4 bytes wont work for everyone.

Reply:
ah, thank you for reminding. the generic version does not require this limit, so maybe in this way would be better.
`#if defined(__XCC__) && XCHAL_HAVE_HIFI3
	/* Get aligned source, sink, number of frames etc. to process. */
	comp_get_copy_limits_with_lock_frame_aligned(source, sink, &c);
#else
	comp_get_copy_limits_with_lock(source, sink, &c);
#endif`

In the implementation of hifi version, we use multi-operation for each format (2-way for S24 & S32, 4-way for S16 ) of volume component, so we must process multiple of 4 frames to avoid unexpected address hopping. Since each sample size would be 2-byte or 4-byte, if frames is a multiple of 4, it can meet hifi intrinsic 8-byte alignement requirement, too. So whatever channel or format it is, I think the 4 frames aligned would meet all requirement.

`#if defined(__XCC__) && XCHAL_HAVE_HIFI3
	/* volume component process need sink and source frames aligned with 4 */
	sinkb->stream.frame_align_shift = 2;
	sourceb->stream.frame_align_shift = 2;
#endif`

Reply:
I would probably define some static inline functions to get the volume alignments (and pass in the format channels etc) i.e.
```c
#if HIFI3
volume_get_alignment(chan, format)
{ 
code 
}
#elif HIFI4
volume_get_alignment(chan, format)
{ 
code 
}
#else
/* generic C */
volume_get_alignment(chan, format)
{ 
code 
}
```
and volume_get_alignment would be called by prepare().

Reply:
Yep, that would make sense, even if now the alignment would be the same. But later with HiFi5 and Atom those would differ.